### ISSUE TITLE
Increase linkages between SMS reports and other docs to streamline analytics queries

### ISSUE THREAD

**Issue by sglangevin** [medic/medic-webapp#3959]
Currently, SMS reports have limited information attached to them as this can't be adjusted using config the way it can be for app forms. This causes challenges on the analytics side of things, resulting in complex queries or the inability to calculate certain metrics. This issue will add some linkages between SMS reports and other reports or docs in the DB to help streamline analytics queries.

NOTE: I'm using ANC examples here to illustrate the requests but this should be implemented in a generic way so that it can be used for any use case.

Assumptions:
- N forms are the forms that create new people in the web app
- P forms register pregnancies
- V forms can be any type of visit
- D forms are delivery reports

Linkages to add:
- [x] When an N form is submitted, the person doc that is created should have a `source_id` field with the uuid of the N form that triggered person creation AND a `created_by` field with the uuid of the contact who submitted the N form.
- [x] When a V form is submitted, add a `source_id` field with the uuid of the form that generated the schedule that the V is responding to. For example, if a user sends a P form and then a V form, the V form would have the uuid of the P form on it since the P form generated a schedule.
- [x] When a D form is submitted, add a `source_id` field with the uuid of the form that generated the schedule that the D is responding to. For example, if a user sends a P form, them a V form, then a D form, the D form would have the uuid of the P form.
- [x] Add a `cleared_by` field to any scheduled message that is cleared before it is sent. This field should contain the uuid of the report that cleared the scheduled message.

@michaelkohn is available to provide additional information or talk through scenarios on any of these items.

**Comment by garethbowen** [comment]
@michaelkohn Is it ok with you if we use more descriptive field names? For example, in the first checkbox above `source_id` is a bit generic and something like `registration_report` might be clearer.

Does the old data have to be migrated for this to be useful, or is it ok if it only applies to reports that are received after the instance is upgraded?

**Comment by michaelkohn** [comment]
Thanks for the suggestion.  I don't necessarily love the term `source_id`, but I do like that it works for all use cases... and `source_id` is what is already being used for mobile app forms.  If we went for use case specific naming, it would be worth thinking through what we would call it for the others because they'll need this field as well.

If we go with use case specific naming, we'd probably want to have the field called something like
- registration_uuid
- pregnancy_uuid
- delivery_uuid
- etc...

Though it seems like we would have to code new ones every time we add a new use case.

Another option would be to keep calling it `source_id` (or better, `source_uuid` since thats what it really is) and having a `source_uuid_form` field which tells us the specific form (P, D, etc...) that generated it... that way it also works easily for collect.

One other thing to consider, which I'm not sure will ever be relevant for SMS projects but is worth thinking about... on the mobile app side, if you have an pregnancy (for example) that requires a follow-up, and that follow-up then requires a follow-up, we are able to know the original pregnancy id on the second follow-up (even though the source_id on the second follow-up is actually the first follow-up).  Again... I'm not sure if that is even possible on the SMS side.

RE: Migrations... technically speaking in PostgreSQL we can check for those fields and if they don't exist, we can calculate it, but in the ideal situation, we would update the old docs.

**Comment by garethbowen** [comment]
I decided to use `registration_id` on the visit and delivery forms - it just felt the most useful.

**Comment by garethbowen** [comment]
Code review please @SCdF

**Comment by SCdF** [comment]
Mostly good, back to you

**Comment by sglangevin** [comment]
@michaelkohn flagging this as we'll need to make sure this meets your needs.

**Comment by michaelkohn** [comment]
Cool.  Just to clarify, did it only get done for Visit and Delivery forms and not the others mentioned in the original ticket (ie other use cases, `cleared_by`, etc...)?  Just want to make sure those others are still tracked somewhere if not.

Also... for the visits and deliveries, what does it do if they are not responding to a schedule?  For example... the patient has been registered with an `N` but no `P` is sent (so there is no schedule) but then a `V` or `D` is sent.  Does the field get added but the value is just set to NULL or are we not adding the field at all in that situation?

**Comment by sglangevin** [comment]
The `cleared_by` field was implemented, per the first comment in this thread. I'm not sure about other use cases, but I would assume that this was done generically, so that `patient_reports` will have the uuid of the `registration` that generated the reminders. Delivery forms and other similar forms are a little trickier because they can be both `patient_reports` and `registrations` (using SMS form config lingo), depending on the workflow. Perhaps @garethbowen can clarify though.

**Comment by garethbowen** [comment]
@michaelkohn @sglangevin 

> Delivery forms and other similar forms are a little trickier because they can be both patient_reports and registrations (using SMS form config lingo), depending on the workflow.

Any report that clears a schedule either via the `accept_patient_reports` transition or the `clear_schedule` trigger in the `registration` transition will have the ID of the latest matching registration recorded in the `registration_id` field.

> did it only get done for Visit and Delivery forms and not the others mentioned in the original ticket (ie other use cases, cleared_by, etc...)?

Any report that runs the `registration` transition and invokes the `add_patient` (or deprecated `add_patient_id`) trigger will create a patient doc with a `source_id` field referencing the report.

Any task that is cleared before being sent will have a `cleared_by` field with the report that caused the task to be cleared.

I think this covers all the use cases in this issue.

**Comment by sglangevin** [comment]
I've just tested and if you register a new person via SMS, then send in a V, the `registration_id` is set to the uuid of that `N` form. If you register a new person via the web app, then send in a V, the `registration_id` field doesn't exist.

**Comment by sglangevin** [comment]
@michaelkohn I'm assigning this to you for further AT. Can you make time this week so that we can move this along?

**Comment by garethbowen** [comment]
> If you register a new person via the web app, then send in a V, the registration_id field doesn't exist.

@sglangevin That's strange. Is that on a public instance that I can look at? Can you remember the people you registered? I'd like to have a look!

**Comment by michaelkohn** [comment]
@sglangevin I'll try, but realistically it'll be Tuesday next week.  Do we have test scenarios / acceptance criteria written out somewhere already?  Also, can you let me know _where_ to test?

**Comment by sglangevin** [comment]
@michaelkohn I'm happy to outline what I've tested so far and we can talk more about SMS config if that helps provide useful context to what Gareth implemented.  I used the scenarios described in the issue for testing - goal was to confirm that all are working as expected. It can be tested on alpha.dev which has the Standard config. Here is what I've tested:
- Submit N form using medic reporter. Confirm that the contact generated by this report has a `source_id` that links to the N form and a `created_by` that links to the contact who sent the N form.
- Submit P form then V form using medic reporter. Confirm that the V form has a field called `registration_id` that has the uuid of the most recent P form.
- Submit D form using medic reporter. Confirm that the D form has a field called `registration_id` that has the uuid of the most recent P form.
- Look at the original P form in Futon and review the statuses of the scheduled messages. All should be `cleared` and there should be a `cleared_by` field that refers to either the D form you just sent or the V form you had sent.

The other tests we'll need to do are sending in the different reports out of order and making sure that the results match your expectations. I know these features are meant to help on the analytics side, so I also need to know whether what's there is sufficient for those purposes. Perhaps we can work through this together next week.

@garethbowen I'll get back to you on that today, will probably need to re-test as I'm not sure that I will find the one from before.

**Comment by sglangevin** [comment]
@garethbowen following up - these are on alpha.dev. 

First I registered Jennifer with this [N form](https://alpha.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/a12e2ef85a36226f2e4baddbb2352033) then sent in this [V form](https://alpha.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/a12e2ef85a36226f2e4baddbb2352e45). The V form has a `registration_id` that corresponds to the N form.

Then I registered Martha using the web app form for adding a new person. And then I sent in this [V form](https://alpha.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/a12e2ef85a36226f2e4baddbb2353e41). The V form doesn't have the `registration_id` field.

Hope that helps!

**Comment by garethbowen** [comment]
@sglangevin Sorry for the delay here. I've managed to have a look and it's just as you say - the `registration_id` is only set when the patient shortcode is found on a registration form and not on a contact. It might be quite confusing if it sometimes linked to a report and sometimes to a contact but it probably depends how it's used. Alternatively we could consider having a different field for `patient_id` which links to the contact doc. It comes down to how the contact doc will be used.

@michaelkohn What do you think?

**Comment by michaelkohn** [comment]
Attached some of the test scenarios I went through.

These worked as expected
- The `cleared_by` functionality in schedules
- The `source_id` field on  `person` records
- The `created_by` field on `person` records

These did not work as expected 
- Submit N and then delivery.  The `registration_id` on the `delivery` linked to the `N`
- Submit P then P then P.  The `registration_id` on subsequent `P`'s linked to the previous `P`
- Submit D then delivery.  The `registration_id` on the `delivery` linked to the `D`

Some general thoughts/comments (using ANC as an example)
- We don't need any linkage IDs on pregnancies (this wasn't on the original issue), we can link them via the patient id.
- The linkage ID on a `V` should either be a `pregnancy`, `P`, or `NULL`.  We never need to link it back to the `N` form (since we'll have the `patient_id`).  
- When calculating the linkage ID on a `V`, we should also have rules to 'how far back' we look for pregnancies.  For example... if the most recent pregnancy was 2 years ago, we shouldn't link this `V` to that `pregnancy`.  Perhaps we could say only look for pregnancies that have an open schedule. 
- The rules for the `D` should be similar to the `V`... the linkage ID should only refer to a `pregnancy`, `P`, or be `NULL`.  I'm not sure how far out our pregnancy schedules typically go (i.e. how far beyond the EDD), but we'd want to do something similar to what I mentioned above whereby we don't  link the `D` to a `P` from 2 years ago.  If there isn't a recent `P`, this `D` is a free agent (i.e. the linkage ID should be NULL)


[3959 testing.txt](https://github.com/medic/medic-webapp/files/1791443/3959.testing.txt)

**Comment by sglangevin** [comment]
@garethbowen sounds like it is not optimal for a report to be linked to a contact in that way. I think it should always be linked to a report.

@michaelkohn thanks for doing all that testing. Here is my feedback:
> We don't need any linkage IDs on pregnancies (this wasn't on the original issue), we can link them via the patient id.

I'm not sure we can avoid having linkage IDs on pregnancies if we need them on delivery forms because both of these forms are registrations. Can't you just ignore them for pregnancy forms? @garethbowen do you have any ideas for how we could determine which forms should and should not have linkage IDs? Perhaps we can leave this one for now and improve upon it later. 

> The linkage ID on a V should either be a pregnancy, P, or NULL. We never need to link it back to the N form (since we'll have the patient_id).

I think this would be solved by the comment at the top - never linking reports to people/people creation forms, but I'm not sure. We could probably make it so that we couldn't link to a form that is used solely for creating people, but that might be tricky.

> When calculating the linkage ID on a V, we should also have rules to 'how far back' we look for pregnancies. For example... if the most recent pregnancy was 2 years ago, we shouldn't link this V to that pregnancy. Perhaps we could say only look for pregnancies that have an open schedule.

@garethbowen is it possible to narrow to registrations with an open schedule? That would avoid linking to very old pregnancies that are likely irrelevant.

> The rules for the D should be similar to the V... the linkage ID should only refer to a pregnancy, P, or be NULL. I'm not sure how far out our pregnancy schedules typically go (i.e. how far beyond the EDD), but we'd want to do something similar to what I mentioned above whereby we don't link the D to a P from 2 years ago. If there isn't a recent P, this D is a free agent (i.e. the linkage ID should be NULL)

I think this would be solved by the previous comment. Also, there are cases where the registration ID field isn't generated at all - do we need to make it so that it is always generated and only filled in if an appropriate form is identified?

**Comment by garethbowen** [comment]
> Submit N and then delivery. The registration_id on the delivery linked to the N

As discussed, this is supporting the 1-step registration. Because there is no `R` there may actually be useful information (eg: lmp_date) on the `N` form. I think this is the correct behaviour.

> Submit P then P then P. The registration_id on subsequent P's linked to the previous P

Definitely a bug. I'm working on a patch for that now.

>  Is it possible to narrow to registrations with an open schedule? That would avoid linking to very old pregnancies that are likely irrelevant.

I think this should be done as a validation which, if found to be invalid, would halt the transition before the `registration_id` was added. As far as I can tell it's difficult, if not impossible, to write a validation that checks for a particular form within a time frame. @sglangevin Maybe we should raise a new feature request for this? It seems like a very useful feature...

**Comment by garethbowen** [comment]
Code review please @SCdF

**Comment by sglangevin** [comment]
Assigning to @michaelkohn for AT.

@garethbowen it does seem like a potentially useful feature - do we have other use cases for it or is it mostly an improvement for analytics when linking reports?

**Comment by garethbowen** [comment]
It's a general validation improvement - if a project doesn't want to allow someone to do an ANC visit for a 2 year old pregnancy this would allow the message to get rejected. This could protect from accidentally entering the wrong pregnancy number but also if you've forgotten to register the new pregnancy.

**Comment by sglangevin** [comment]
Interesting - so it would allow us to do something like rejecting a V form if there was no pregnancy registered in whatever time period we set and send an auto-response that says: please register a pregnancy before reporting a visit?

**Comment by garethbowen** [comment]
Exactly! We have validation functions for

- `unique`: check whether another form of the same type exists with the same configurable fields
- `uniqueWithin`: same as above but only after a certain date
- `exists`: check whether a form of a different type exists with a fields set to a certain value

One for `existsWithin` might be enough because we could check for a registration within the last 9 months or something. Or we could write a really sophisticated one which also checks for delivery forms.

**Comment by michaelkohn** [comment]
Testing on https://alpha.dev.medicmobile.org/

- I sent this [N](http://localhost:5984/_utils/document.html?medic/0a976c4dcc9a90c81ad2d5f4ce03660e) and it created this [person](http://localhost:5984/_utils/document.html?medic/0a976c4dcc9a90c81ad2d5f4ce03759d).  There was no `source_id` on the `person`.

- Sent this [P](http://localhost:5984/_utils/document.html?medic/0a976c4dcc9a90c81ad2d5f4ce03833a) and there was no `registration_id` on that P; However, there was an error **missing_named_view** but I don't really know what this means or how to resolve it.

**Comment by sglangevin** [comment]
I'd test on beta.dev because alpha.dev is now on version 3.0

**Comment by michaelkohn** [comment]
[Messages](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/messages/contact:d31881ca-d3c7-4b0b-8188-008fe75b159d) from Medic Reporter are not being recognized as forms on beta.dev.  I sent it from [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/d31881ca-d3c7-4b0b-8188-008fe75b159d) CHW.



<img width="1199" alt="screen shot 2018-03-23 at 8 23 58 am" src="https://user-images.githubusercontent.com/10096985/37808640-93395ffc-2e73-11e8-99b5-1f3c41ab7ba0.png">

**Comment by sglangevin** [comment]
Shoot, sorry - I forgot that doesn't have the Standard config on it. I can fix that for you tonight.

**Comment by michaelkohn** [comment]
> do we have other use cases for it or is it mostly an improvement for analytics when linking reports?

As Gareth mentioned... it's a general improvement, you'll have the same issue on contact summary whereby new visits might get associated to old pregnancies.  I've already seen things get wonky when you have multiple pregnancy registrations and only one delivery.... the Past Pregnancies card shows that the woman delivered multiple times (one for each pregnancy registered).

***
> so it would allow us to do something like rejecting a V form if there was no pregnancy registered in whatever time period we set and send an auto-response that says: please register a pregnancy before reporting a visit?

We should try to coordinate the rolling out of the `registration_id` linkages with the inclusion of some of these validations.  If not, I'll always have to first check to see that the linkage is valid - that it isn't linking to an older P (for example) - before relying on it.  Technically I can do that, but it's bad practice to have the wrong value in that field in the first place.

I know this has wide reaching workflow implications and may be a bit off topic, but here's my "wishlist towards data integrity", using ANC/PNC as an example.

- Do not allow a new pregnancy if there is already an existing active pregnancy (or automatically archive the previous one)
- Do not allow pregnancy visits if there is no pregnancy
- Do not allow a delivery without a pregnancy (or automatically generate the pregnancy)
- Do not allow a delivery if there is already an existing recent delivery (or automatically archive the earlier one)
- Do not allow PNC visits if there is no delivery
- Do not allow F if there is no active pregnancy or delivery for the woman
- Do not allow deletion of a pregnancy without first deleting associated pregnancy visits
- Do not allow deletion of a delivery without first deleting associated pnc visits

**Comment by SCdF** [comment]
> Sent this P and there was no registration_id on that P; However, there was an error missing_named_view but I don't really know what this means or how to resolve it.

That should never happen, that means someone has misspelled a view or something like that. It's 99% a bug in our code. Was there more in the logs?

**Comment by henokgetachew** [comment]
@SCdF this might be related to [this](https://github.com/medic/medic-webapp/issues/4326#issuecomment-376180643).

**Comment by ngaruko** [comment]
@michaelkohn are you able to do a final check and confirm that the feature/issue works as you expect (and eventually close it if satisfied)? Thanks.

**Comment by michaelkohn** [comment]
Has the `missing_named_view`issue been resolved?  And if so, can you please let me know which instance this is configured on so that I can test.

**Comment by ngaruko** [comment]
The `missing_named_view` issue #4326 was closed because it could not be reproduced. I guess `beta.dev` has now the right config for this test. Aye @sglangevin ?

**Comment by michaelkohn** [comment]
Testing on https://beta.dev.medicmobile.org

## SMS workflow testing
- I sent this [N](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5a5298b0a308146efa4ab51d74690f13) through medic reporter and it created this [person](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/5a5298b0a308146efa4ab51d74690fd9) with a `patient_id = 11459`.  I was able to find the `source_id` on that record in [futon](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d74690fd9) and it correctly links to the `N`.  ***This test passes***.

- Sent this [P](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5a5298b0a308146efa4ab51d74691af9) through medic reporter and there was no `registration_id` on this record, which is the desired behavior.  This form can be found [here](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d74691af9) in futon.  ***This test passes***.

- Sent another [P](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5a5298b0a308146efa4ab51d7469262d) through medic reporter.  This cleared all messages on the previous  [P](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d74691af9) and the `cleared_by` key was added and correct.  The new `P` also did not have the `registration_id` field, which is the desired behavior.  ***This test passes***.

- Sent this [V](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5a5298b0a308146efa4ab51d74692a40) through medic reporter.  The `registration_id` of the `P` was added to the record, was correct, and can be viewed in futon [here](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d74692a40).  ***This test passes***.

- Sent this [D](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/5a5298b0a308146efa4ab51d7469372e) through medic reporter.  There was no `registration_id` on this `D` as seen in futon [here](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d7469372e).  In addition, the `cleared_by` was successfully applied on the [P](http://localhost:5984/_utils/document.html?medic/5a5298b0a308146efa4ab51d7469262d). ***This test passes***.

## App testing
- Created this [person](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/209d0c87-e7ef-4f68-a341-4c294b8726c7) through the app. This did not add the `source_id` as seen [here](http://localhost:5984/_utils/document.html?medic/209d0c87-e7ef-4f68-a341-4c294b8726c7) in futon, which is the expected behavior. ***This test passes***.

- Submitted this [pregnancy](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/d6daf04c-bcbb-4461-a20d-1cd85de593b5) from the app.  The `registration_id` was not added to this form which can be seen [here](http://localhost:5984/_utils/document.html?medic/d6daf04c-bcbb-4461-a20d-1cd85de593b5) in futon, this is the expected behavior.  ***This test passes***.

- Submitted this [pregnancy_visit](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/464a3f80-74fa-4a59-bf0c-3a6c99c1b95e) from the app.  As seen [here](http://localhost:5984/_utils/document.html?medic/464a3f80-74fa-4a59-bf0c-3a6c99c1b95e) in futon, there is no `registration_id` on this pregnancy_visit.  ***This test fails***.

- Submitted this [delivery](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/0091ab70-198d-4dbf-9848-7000f3bad9d2) from the app.  The `cleared_by` was added to the `pregnancy` record and was correct.  ***This test passes***.

## Unrelated issues

- The Past Pregnancies condition card is incorrect when there are multiple pregnancies.  Not only does it duplicate the number of deliveries, the label is also wrong (I'm assuming it is supposed to be PNC visits at this point).

<img width="710" alt="screen shot 2018-04-09 at 2 06 08 pm" src="https://user-images.githubusercontent.com/10096985/38487229-31f7bfe2-3bff-11e8-8474-5caced606502.png">

**Comment by abbyad** [comment]
> ## Unrelated issues
> The Past Pregnancies condition card is incorrect when there are multiple pregnancies. Not only does it duplicate the number of deliveries, the label is also wrong (I'm assuming it is supposed to be PNC visits at this point).

This appears to be related to the Standard config... @michaelkohn can you open an issue for this in medic-projects? @jillrshah (and I) will follow up on it.

**Comment by abbyad** [comment]
Moving back to _Active_ due to the failed test.

**Comment by garethbowen** [comment]
The failed case is because the `accept_patient_reports` transition didn't run which is because the senders parent is a health_center not a clinic.

For some reason the transition [checks that the sender has a clinic that has a phone](https://github.com/medic/medic-webapp/blob/master/sentinel/transitions/accept_patient_reports.js#L189). I think the right thing to do is remove this check so the transition will run on nurse submitted docs and also for clinics without phones to make it more useful for non-sms instances.

@abbyad Does this make sense to you?

**Comment by abbyad** [comment]
Yes, makes sense to me as it doesn't matter _who_ submits the report. Not sure what side effects changing this will have though.

**Comment by garethbowen** [comment]
Code review please @rmhowe . This needs to be backported to `2.15.x` but feel free to leave that to me...

**Comment by abbyad** [comment]
@michaelkohn what's the latest on AT for this issue?

**Comment by michaelkohn** [comment]
Testing on https://beta.dev.medicmobile.org

Created [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/7db9c5e2-1231-4f1c-b22d-55da35091b5a) person from the webapp.   I was a bit surprised that the Medic ID was not assigned.

Then I submitted [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/7e6beb0d-05a5-4136-b733-bbb0e1bfe668) pregnancy from the webapp.  The pregnancy does not show up under the person I created.

No sense in continuing testing since the pregnancy_visit relies on the pregnancy, and that part is already failing.

<img width="1176" alt="screen shot 2018-04-17 at 8 51 53 am" src="https://user-images.githubusercontent.com/10096985/38846541-a269601e-421c-11e8-8623-6a0c0ae28ca0.png">

<img width="1192" alt="screen shot 2018-04-17 at 8 52 03 am" src="https://user-images.githubusercontent.com/10096985/38846543-a469cbb0-421c-11e8-8f77-f2a2c5bfabf2.png">

**Comment by michaelkohn** [comment]
FWIW,  I'm able to replicate the issue.  Here's a video

**Comment by abbyad** [comment]
The patient ID normally shows up a couple of seconds after the person was created, perhaps took longer on your side so wasn't available by the time you opened the form. 

![image](https://user-images.githubusercontent.com/789512/38889087-9243a04e-424b-11e8-870f-9d81b3ad12fa.png)

That said, **the form should still use the UUID for the person and show up in the right place.**

**Comment by abbyad** [comment]
@michaelkohn in the meantime can you proceed with AT while we look at what's going on with the missing ID? [Here is a pregnancy report](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/f1943599-49cd-4835-a4a3-c458c1030952) similar to the one in your video that may work for AT.

**Comment by abbyad** [comment]
I can reproduce the scenario that @michaelkohn found, will try to distill a bit more and create a new issue for it if needed.

**Comment by michaelkohn** [comment]
Created [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/contacts/2f74054d-c5c6-430f-90dc-5f25b6d2dec2)  `person` from the webapp.

Submitted [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/03fd5e71-9688-4e4d-8722-d587d16bc637) `pregnancy` from the webapp.

Submitted [this](https://beta.dev.medicmobile.org/medic/_design/medic/_rewrite/#/reports/3819ff27-66c2-4c80-ae97-1132f599b8ad) `pregnancy_visit` from the webapp and viewed [this](http://localhost:5984/_utils/document.html?medic/3819ff27-66c2-4c80-ae97-1132f599b8ad) record in futon.  The record correctly has the `registration_id` added.

I haven't gone back and re-tested all the original scenarios, but the most recent failure... 
>Submitted this pregnancy_visit from the app. As seen in futon, there is no registration_id on this pregnancy_visit. This test fails.

...has now been fixed.

**Comment by michaelkohn** [comment]
Yep!
